ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/manager/bounded/BoundedTimedQueue.java

```diff
@@ -0,0 +1,163 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.eth.manager.bounded;
+
+import static java.util.Objects.requireNonNull;
+
+import tech.pegasys.pantheon.metrics.Counter;
+import tech.pegasys.pantheon.metrics.MetricsSystem;
+import tech.pegasys.pantheon.metrics.PantheonMetricCategory;
+import tech.pegasys.pantheon.metrics.noop.NoOpMetricsSystem;
+
+import java.time.temporal.ChronoUnit;
+import java.time.temporal.TemporalUnit;
+import java.util.Locale;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class BoundedTimedQueue extends LinkedBlockingDeque<Runnable> {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final long keepAlive;
+  private final TemporalUnit keepAliveUnit;
+  private final long evictionPollTimeout;
+  private final TimeUnit evictionPollTimeoutUnit;
+  private final MetricsSystem metricsSystem;
+  private final String metricName;
+  private final Counter totalEvictedTaskCounter;
+
+  public BoundedTimedQueue() {
+    this(
+        Default.CAPACITY,
+        Default.KEEP_ALIVE,
+        Default.KEEP_ALIVE_UNIT,
+        Default.EVICTION_POLL_TIMEOUT,
+        Default.EVICTION_POLL_TIMEOUT_UNIT,
+        Default.METRICS_SYSTEM);
+  }
+
+  public BoundedTimedQueue(
+      final int capacity,
+      final long keepAlive,
+      final TemporalUnit keepAliveUnit,
+      final long evictionPollTimeout,
+      final TimeUnit evictionPollTimeoutUnit,
+      final MetricsSystem metricsSystem) {
+    super(capacity);
+    this.keepAlive = keepAlive;
+    this.keepAliveUnit = keepAliveUnit;
+    this.evictionPollTimeout = evictionPollTimeout;
+    this.evictionPollTimeoutUnit = evictionPollTimeoutUnit;
+    this.metricsSystem = metricsSystem;
+    this.metricName = this.getClass().getSimpleName().toLowerCase(Locale.US).replace('-', '_');
+    this.totalEvictedTaskCounter =
+        this.metricsSystem.createCounter(
+            PantheonMetricCategory.EXECUTORS,
+            metricName + "_rejected_tasks_total",
+            "Total number of tasks rejected by this working queue.");
+  }
+
+  @Override
+  public boolean offer(final Runnable runnable) {
+    try {
+      requireNonNull(runnable);
+      if (!(runnable instanceof TimedTask)) {
+        return false;
+      }
+      final TimedTask task = (TimedTask) runnable;
+      if (!isEligible(task)) {
+        LOG.debug("Task does not match time policy, rejecting task.");
+        return false;
+      }
+      final boolean offered = super.offer(task);
+      if (offered) {
+        return true;
+      }
+      // Try to evict tasks based on time policy.
```

Comment:
So, we'll only start evicting when we hit 1,000,000 items?  Not sure how well that will work out in practice - seems like at that point we might be overwhelmed?  And then we may end up blocking for a while as we empty out the task queue.

Reply:
I see your point. However how to decide the number of messages to drop if we remove the time criteria. I mean when there is no more space in the queue the current eviction process remove tasks from the head until a non expired task is found. Should we use a fixed number of messages to drop ?

Reply:
Discussed internally. We decided to remove the time based policy. We make a room for a new element when we reach max capacity.